### PR TITLE
feat: Add provisioner label to interrupt metrics

### LINKED ISSUE #3466: Add provisioner to karpenter_deprovisioning_actions_performed metric
### Version

__Karpenter Version:__ v0.23.0

__Kubernetes Version:__ v1.24.8-eks-ffeb93d

### Expected Behavior

When consolidation is done, we would like to know which provisioner it was done.

### Actual Behavior

The metric `karpenter_deprovisioning_actions_performed` shows that total consolidation of the cluster.the

### Steps to Reproduce the Problem

None

### Resource Specs and Logs

None

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by sbhrule15**
Fixes #3466 

**Description**

Adds an additional `provisioner` label to the following metrics to allow grouping actions by provisioner:
karpenter_interruption_received_messages
karpenter_interruption_actions_performed

**How was this change tested?**

This change was tested both through the automated tests already in the repository, as well as multiple cycles of deploying to an EKS cluster with the inflating and deflating of pause images to simulate load, between zero capacity, full capacity, middle capacities, and cycling between combinations of the three.

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note
Adds `provisioner` label to interruption and deprovisioning metrics
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/interruption/controller.go [hunk 1]
```diff
@@ -155,6 +158,10 @@ func (c *Controller) handleMessage(ctx context.Context, machineInstanceIDMap map
 			continue
 		}
 		node := nodeInstanceIDMap[instanceID]
+		receivedMessages.With(prometheus.Labels{
```
Comment (jonathan-innis) [inline pkg/controllers/interruption/controller.go:161]: What happens if the message has multiple instance IDs that are part of different provisioners?

File: pkg/controllers/interruption/metrics.go [hunk 2]
```diff
@@ -34,9 +35,9 @@ var (
 			Namespace: metrics.Namespace,
 			Subsystem: interruptionSubsystem,
 			Name:      "received_messages",
-			Help:      "Count of messages received from the SQS queue. Broken down by message type and whether the message was actionable.",
```
Comment (jonathan-innis) [inline pkg/controllers/interruption/metrics.go:37]: I'm thinking the only metric here that _really_ makes sense to add the provisioner label is the `actionsPerformed` metric. Every other metric has to deal with SQS messages generally so we don't really have context-specific information about the provisioner in there.

**Review by jonathan-innis — COMMENTED** [review]
Looks good 🎉 Just a couple comments

File: pkg/controllers/interruption/metrics.go (hunk 2 above)
Comment (sbhrule15) [reply pkg/controllers/interruption/metrics.go:37]: That makes sense, I agree - I'll remove

File: pkg/controllers/interruption/controller.go (hunk 1 above)
Comment (sbhrule15) [reply pkg/controllers/interruption/controller.go:161]: Wouldn't that be handled from this being in the loop of all the instance ID's in the message, and looking up the provisioner by using the label on the machine directly for each instance ID?

File: pkg/controllers/interruption/controller.go [hunk 3]
```diff
@@ -144,9 +144,11 @@ func (c *Controller) handleMessage(ctx context.Context, machineInstanceIDMap map
 	nodeInstanceIDMap map[string]*v1.Node, msg messages.Message) (err error) {
 
 	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("messageKind", msg.Kind()))
-	receivedMessages.WithLabelValues(string(msg.Kind())).Inc()
 
 	if msg.Kind() == messages.NoOpKind {
+		receivedMessages.With(prometheus.Labels{
```
Comment (jonathan-innis) [inline pkg/controllers/interruption/controller.go:149]: I'm not sure that anything needs to be changed in how we are currently handling the `receivedMessages` metric. This metric would fire more than needed now with the changes you made. I think we can probably ignore this metric for this set of changes.

**Review by jonathan-innis — CHANGES_REQUESTED** [review]

**Review by jonathan-innis — APPROVED** [review]
/karpenter snapshot

**Comment by jonathan-innis** [comment]
@sbhrule15 Just looks like you need to run `make presubmit` to fix the CI failures. Otherwise, looks good 🎉

**Comment by jonathan-innis** [comment]
Closing this one since #3466 is closed.

**Pull request closed without merging**